### ISSUE TITLE
[BUG] Some cookies appear to be stuck / CAD says it was deleted when it is not  (FF 86+ Strict Mode / cross-site protection / dynamic first party isolation / total cookie protection) 

### ISSUE THREAD

**Issue by neroux** [mrdokenny/Cookie-AutoDelete#807]
**Describe the bug**

It appears as if under certain conditions CAD was not able to remove eligible cookies on **Firefox**, even though it tries. Even though it appears consistently in my browser profile I found it difficult to exactly reproduce it in a new browser profile. Cookies got stuck there too, but not the same.

**To Reproduce**

1. Go to a Google site (google.com, youtube.com). Had it with Wikimedia cookies however too.
2. Make sure you got cookies set.
3. Close the tab.
4. CAD should now attempt to clear the cookies in question.

**Expected behavior**

All eligible cookies should get removed.

**Screenshots**

Three attempts over four seconds

![image](https://user-images.githubusercontent.com/24754773/88113371-f4818400-cbb1-11ea-987b-c2b6f2b46ee3.png)

That screenshot should hopefully clearly show that CAD does determine these cookies as eligible for removal, but still fails to remove them

**Your System Info**

 - OS: Windows 8.1
 - Browser Info: Firefox 80
 - CookieAutoDelete Version: 3.4.0

**Additional context**

To be fair, it could be that is a general Firefox issue. With https://addons.mozilla.org/firefox/addon/cookie-quick-manager/ I don't seem to be able to delete the cookies selectively either, I need to use Firefox own built-in cookie manager.

The cookies on "accounts.youtube.com" are called `CheckConnectionTempCookie`. Are there any known issues in this area?

**Comment by neroux** [comment]
> Browser Info: Firefox 80

It might be some specific to the new version. On 78 I do not seem to be able to reproduce it (cookies do get properly removed).

To rule out other extensions, I now also disabled all other extensions and cookies still remained.

I can't tell how CAD removes cookies, but from a log excerpt from aforementioned cookie manager, it would seem as if [`browser.cookies.remove()`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/remove#Return_value) returns `null` for some reason

> If a cookie matching the details parameter could not be found, the promise is fulfilled with null.

**Comment by kennethtran93** [comment]
Normally I'd probably wait until it at least wait until Firefox Dev Edition reaches that version before starting support for issues. With Nightly, some things may break at times. 

Do you have first party isolation and firstparty...use_site enabled? 

That's the cause and I found the browser changes that made it happen (at least in 78). There's a recent issue that I commented in that reflects the recent changes in regards to use_site setting. (#801)

**Comment by neroux** [comment]
Thank you for the quick response. Fair enough, Mozilla often breaks things, so it probably is reasonable to wait until the first beta/developer edition, which should be next Monday, the 27th.

Let me please check about the two settings you asked, though I **believe** I haven't changed the defaults here. I'll get back to you shortly

**Comment by neroux** [comment]
> Let me please check about the two settings you asked, though I believe I haven't changed the defaults here. I'll get back to you shortly

I just checked on another machine (where the issue appears as well with 80.0) where I have the following settings

![image](https://user-images.githubusercontent.com/24754773/88186293-491a1300-cc35-11ea-9212-e1e9b6bfef55.png)

Will check on the original machine later too.

**Comment by neroux** [comment]
> Will check on the original machine later too.

Just checked, same settings as above.

**Comment by kennethtran93** [comment]
I won't be looking into this until it hits Dev Edition, but if you want to help diagnose this, you can try and see what changes were made that altered the webextension api for cookies, notably the ones for cookies.getAll and cookies.remove.  It is weird as we pass in the cookies returned from cookies.getAll marked for removal so it should not return null at all...

**Comment by neroux** [comment]
As I said, that is fair enough, however 80.0 will hit beta next Monday already anyhow :)

I cannot tell if it returns `null` in your case too, this was just an observation from aforementioned cookie editor (which presumably also takes the cookie object returned by the browser).

It could be that it is just a glitch in Firefox, but again, that version should be beta in less than a week.

**Comment by neroux** [comment]
Just ran it on the current beta, where it still seems to work.

So let's wait until Monday when 80.0 becomes beta and - assuming that change actually makes it into the beta - then we might know more.

**Comment by neroux** [comment]
CAD on the the currently Nightly still exhibits that behaviour. Today 80.0 should be in beta, will check with the Beta too.

Currently I am wondering if it could be related to third party cookies 🤔

**Comment by kennethtran93** [comment]
The interesting thing is that we first call browser.cookies.getAll to get all cookies of each container, then process our filters on it.  If it needs to be cleaned then we call browser.cookies.remove on that cookie, so in reality it should work.

**Comment by neroux** [comment]
I am not particularly familiar with that area of the `browser` object, but I did assume you'd work directly with the cookie objects managed by the browser (equally I believe aforementioned cookie editor does the same).

The third party cookies really is just a wild guess at this moment, it just would seem as if mostly cookies were left over which belong to hosts which were only indirectly visited.

I guess with tomorrow's new beta we should be able to tell if that bug/feature/glitch made it into the beta as well.

**Comment by neroux** [comment]
Mozilla still seems to roll out the new build and so far there is only one Win64 build available at http://ftp.mozilla.org/pub/firefox/candidates/80.0b1-candidates/build1/win64/en-US/Firefox%20Setup%2080.0b1.exe. With that version however I could not reproduce it for now. Possibly the change has not made into the beta yet.

Will try later today the "official" 80.0 beta as well as the new 81.0 Nightly.

**Comment by neroux** [comment]
The new Nightly for 81.0 seems to exhibit that issue as well.

**Comment by neroux** [comment]
All right, quick summary. CAD on

* 80 Nightly exhibited the issue
* 80 Beta does not exhibit it any longer
* 81 Nightly is exhibiting it once more

It seems they introduced some change in the Nightly recently which was not merged any more to the beta branch. Whether that will land in the next beta is naturally something I cannot tell. 81 Beta is scheduled for August 24.

**Comment by kennethtran93** [comment]
Now that FF 80.0b1 is on Dev channel, I attempted to use a fresh profile but couldn't reproduce the cookie issue.  Unfortunately because I don't have your profile it is hard to say.  If it works fine on a fresh profile then I may have to just ask you to use a fresh one.  

Also I ran the test with the 3.5.0 alpha build of CAD which includes the ability to clean cache, indexeddb, plugindata, serviceworkers, in addition to cookies and localstorage.  Could you try that version please?

**Comment by neroux** [comment]
> Now that FF 80.0b1 is on Dev channel, I attempted to use a fresh profile but couldn't reproduce the cookie issue. Unfortunately because I don't have your profile it is hard to say. If it works fine on a fresh profile then I may have to just ask you to use a fresh one.

Ehm, did you read my last three responses?

**Comment by kennethtran93** [comment]
I just wasn't quite sure if the beta you were talking about is the Developer Edition of Firefox or not.

I also believe that I stated earlier on that I would start to look into it once the Developer Edition of Firefox hits 80.

**Comment by kennethtran93** [comment]
However, if you're only using the Beta/Nightly and not Stable/Developer Edition at this time, then I will need to close this as won't fix due to the unstableness of beta and nightly.  I've only kept this open this first time to see if it propagated up to developer edition.

If it does happen in developer edition, then an issue should be opened.

**Comment by neroux** [comment]
Beta and Developer are the same code base.

At this point the issue is rather an FYI as it might show up in the next beta on mentioned date, unless of course Mozilla fixes it by then.

**Comment by kennethtran93** [comment]
Closing as too early of a build to take any action for.

Feel free to revive this or start a new one if it can be reproduced in beta/Dev Edition.

**Comment by neroux** [comment]
I am about to switch to the public channel, so I will probably not follow this issue any more I am afraid.

**Comment by neroux** [comment]
@kennethtran93, can't fully confirm yet, but it might be as if that issue is back in beta 86. Last time this was in version 80 so it might be for a different reason this time, but some cookies still appear to get stuck and removing then returns `null` as previously pointed out.

And yes, this is not about the nightly this time but about the beta.

**Comment by kennethtran93** [comment]
Hmmm. Maybe irrelevant, but what is your firstparty isolation setting at?

**Comment by kennethtran93** [comment]
The only item that I can think of is where we had to generate the url for each cookie. Everything else appears to be straight off the cookie itself.

**Comment by neroux** [comment]
I assume you mean `privacy.firstparty.isolate`, which would be the default false.

**Comment by neroux** [comment]
As I mentioned in the very first posting, this is most likely not a CAD issue but might be an issue introduced by Firefox. Was most likely back then and might be now as well. https://addons.mozilla.org/firefox/addon/cookie-quick-manager/ has the same issue.

**Comment by neroux** [comment]
> And yes, this is not about the nightly this time but about the beta.

I just updated the beta to the version which will be promoted to release in two days and the issue is still there.

Still can't tell whether it's a Firefox or a CAD issue (I'd assume the former) but CAD won't be working properly as of in two days.

**Comment by kennethtran93** [comment]
Things to enhance:
- Start using Promise.allsettled to allow checking of whether or not cookies were deleted.  Currently we're just pooling a promise of cookie removal and executing all at once using Promise.all, which will simply terminate on first error.  cookie.remove returns the deleted cookie or null if not found, so we could start with that, rather than false notifications that cookies are removed.

Things of note with the new dynamic firstparty:
- The value of `firstpartydomain` in the cookie is simply passed into the removal as well, so unless the removal process requires a different value, it _should_  just work. - `(https,github.com)` vs `github.com`.
- Looks like it would only be applied to the `strict` behaviour, unless I am mistaken.
- From what I read online, it should further limit cross-site tracking.
- There is still no iterables for any other browsing data, so we will continue to set our own cookie so that we can trigger browsingdata cleanups.
- `dynamic_firstparty.use_site` was defaulted to `true`, but didn't fill in any cookie value for `firstpartydomain`
- `firstparty.isolate` was manually set to `true`, removed all data, and after reloading github, `firstpartydomain` = github.com
- with `firstparty.isolate.use_site` set to `true`, along with the above, `firstpartydomain` = (https,github.com)

**Comment by neroux** [comment]
@kennethtran93, the underlying problem is not so much that the feedback is wrong, but rather that the cookie is not deleted.

Again, IMHO this likely is an issue with Firefox and CAD is not the only extension getting `null` on `remove` but it nonetheless renders CAD useless at this point.

**Comment by kennethtran93** [comment]
In a sense, it would further help if we were to actually check the return value of the cookie removed to see if it was actually removed, before compiling a list of cookies removed.

**Comment by neroux** [comment]
Difficult to tell at this point. All I can tell is that the behaviour is the same as reported originally.

Can you reproduce it at this point?

**Comment by kennethtran93** [comment]
after visiting youtube on a fresh profile, I do have some cookies remaining, namely NID from google.ca / google.com, an IDE from doubleclick.

Neither of the cookie editors I had installed could remove the cookie either.

**Comment by neroux** [comment]
Yes, that's exactly the behaviour I described back in July.

Can't tell what it is but CAD will lose most of its functionality as of tomorrow.

**Comment by kennethtran93** [comment]
I feel like there should be a bugzilla report for this already.  I'll try searching for it later.

**Comment by neroux** [comment]
@kennethtran93, shouldn't this issue be reopened?

**Comment by neroux** [comment]
@kennethtran93, thanks, though we are closed again.

**Comment by kennethtran93** [comment]
Yes I know. The template has been adjusted later on, so I was trying to figure out what I needed to adjust.

**Comment by neroux** [comment]
No worries, at least we know the bots are not smart enough take over the world just yet ;)

And closed it again. The bot really doesn't want that fixed :)

**Comment by kennethtran93** [comment]
I search for a specific strings, so that's why it knows when to close.

I was trying to see if I could do this on mobile but can't. I'll see what it hung up on when I'm in my desktop.

**Comment by neroux** [comment]
No worries, thanks. It's always fun to do things like that on mobile operating systems. On the desktop that'll be probably a matter of seconds :)

**Comment by kennethtran93** [comment]
The system info section title got a slight adjustment, as well as a colon after the CAD Version line.

**Comment by candiesdoodle** [comment]
For what its worth, i am on 87.0b3 on windows desktop. With both, privacy.firstparty.isolate enabled and disabled, there are many cookies that are not getting deleted. Including account.google.com
I have highlighted a couple

<img src=https://user-images.githubusercontent.com/20420013/109397519-f8059680-795c-11eb-862b-1f1a839f947e.png width=600>
<img src=https://user-images.githubusercontent.com/20420013/109397491-d99f9b00-795c-11eb-8608-34a428f99318.png width=600>

**Comment by kennethtran93** [comment]
Just a minor note that existing cookies may still retain their previous settings (firstpartydomain  value) after you change config settings, so it is best to manually clear all data before running tests to help debug where exactly the issue is and where it is not.

I believe at this time the firstparty dynamic config is the one to look into.

We had removed the 'privacy' permission so we cannot directly check those browser values. We are currently using a workaround to check if the firstpartydomain value is required, but that is it.

**Comment by Zerwin** [comment]
Hi, I think I have the same bug on Firefox 86.0 live/prod. I have cookies persisting even though CAD tries to delete them.

![image](https://user-images.githubusercontent.com/46265592/109437753-ce18a680-7a26-11eb-8966-562f8956dd41.png)

My firstparty isolation was the same as @neroux had. My tracking protection was on strict till now in Firefox, but same result with Standard. 
Anything I can do to help ?

**Comment by neroux** [comment]
> I think I have the same bug on Firefox 86.0

Yes, as far as I can tell this issue is now present in the public version.

**Comment by candiesdoodle** [comment]
So i have been testing out this and i can't find a pattern. Some really random cookies are getting stuck. I don't have ` privacy.firstparty.isolate` enabled; i don't have containers anymore (but previously had, now removed; restarted the browser etc). I am monitoring using the add-on `Cookie Quick Manager`
Cookies that just cannot be deleted using CAD or even Cookie Quick Manager (pressing the bin icon).
````
account.google.com: __Host-GAPS
.google.com: _GRECAPTCHA
youtube.com: VISITOR_INFO1_LIVE; PREF
.info.trustarc.com: __cf_bm
.nanorep.co: u; 7A86E85F; 81F6201F; bc.visitor_token
````
there are some more which have accumulated as I carried on my browsing. Browser version 87b3 and 87b4
As you can see, there is no pattern here. Only way to get rid of these is to delete from the Firefox data manager under options.

**Comment by neroux** [comment]
@candiesdoodle, your observations are identical to mine and what I reported for the Nightly of Firefox 80 in July.

I couldn't determine a pattern either but some cookies simply are stuck for good and trying to remove them from an extension context returns `null`. Only the preferences dialog can remove them.

**Comment by neroux** [comment]
I am pretty sure this is the same issue as reported in July, so that change must have been made back then.

**Comment by maltalex** [comment]
Came in to report that I'm experiencing the same issue on 86.0.1. What tipped me off is that the number of "deleted" cookies in the popup kept growing and I was seeing the same cookies over and over.

**Comment by candiesdoodle** [comment]
I am playing around wit FF developer portable apps version from portableapps.com
It doesn't have any issue related to this 
I have FPI, Temp containers, multi account containers, CAD and a bunch of other things enabled. But still not not seeing any issue related to non deletable cookies

**Comment by kennethtran93** [comment]
@candiesdoodle it really depends on the configuration - namely the firstpartyisolation items as well as the use_site values, in the config page (about:config), or the 'strict' setting in the privacy settings page (about:preferences#privacy).

**Comment by neroux** [comment]
@kennethtran93, you mentioned a Bugzilla bug. Is there any update on this issue? Will there be a fix any time soon? Thanks.

**Comment by maltalex** [comment]
@kennethtran93 FWIW, I do indeed use the *strict* privacy setting. I suspect that many CAD users do. Could you link the Bugzilla issue?

**Comment by kennethtran93** [comment]
@neroux @maltalex I recall reading about a bug before, but took me a while to find it again.

I believe this would be it, ~not sure if it was already linked in other issues already~:  https://bugzilla.mozilla.org/show_bug.cgi?id=1669716 initially noted from issue below.

It would appear this is the patch for the new type of protection with strict mode for 86+: https://bugzilla.mozilla.org/show_bug.cgi?id=1686296

This seemingly talks about migration from FPI to dFPI, though unsure if it will go through: https://bugzilla.mozilla.org/show_bug.cgi?id=1649876

Looks like we had an even earlier issue #719 which may be tied into this already, as this bug report https://bugzilla.mozilla.org/show_bug.cgi?id=1388873#c2 has references which ultimately created the first bugzilla linked above.  I finally recall that's where I read about the bug in the first place.


So the culprit is definitely the dFPI (Dynamic First Party Isolation), which is enabled in strict mode.

On a sidenote, I find it funny that the versions 68 and 86 are both issues that affects CAD in Firefox, former being the mobile version, latter being desktop/dFPI.

**Comment by maltalex** [comment]
@kennethtran93 Thank you for the detailed response (and for your work on CAD!).

**Comment by neroux** [comment]
@kennethtran93, is that something you can fix or are you waiting for Mozilla on this?

**Comment by kennethtran93** [comment]
@neroux unfortunately that is a different programming language I have yet to dig into at this time.  Rob Wu (the one that created cookie manager) seems to already handle a few webextension bits and pieces, and last I read somewhere he may be able to find time in the near future to get a patch going.

**Comment by neroux** [comment]
@kennethtran93, by "you" I meant in the context of CAD. What's not clear to me is if this is something Mozilla needs to fix or the extensions.

From your statement about Rob Wu and the extension I would understand that's something to be fixed on the extension side, so CAD, right?

**Comment by kennethtran93** [comment]
@neroux This is something Mozilla needs to fix, and Rob Wu is part of the Mozilla Firefox WebExtension team, or at least he has contributed to that alongside his webextension.

Long story short, Mozilla introduced another internal attribute/variable for the dynamic first party isolation that the front-facing webextension API has no way of fetching at this time.

**Comment by neroux** [comment]
All right, so it has been a Firefox issue all along then.

**Comment by jabhq** [comment]
Just discovered this thread. Too late to help except to say I became aware of getting continuous notification of around 219 (irrelevant) cookies deleted. Checked size of the cookie storage in Ff and found it to be 0.7GB. I reverted to Firefox handling cookies.

**Comment by modelfe** [comment]
This is also occurring with Custom mode in Firefox 87 when "Cross-site cookies - includes social media cookies" is selected (this option is distinct from "Cross-site _tracking_ cookies - includes social media cookies" and only shows up if you select Strict mode before switching to Custom for some reason). I assume this is the dynamic isolation option in Custom.

**Comment by kennethtran93** [comment]
> This is also occurring with Custom mode in Firefox 87 when "Cross-site cookies - includes social media cookies" is selected (this option is distinct from "Cross-site _tracking_ cookies - includes social media cookies" and only shows up if you select Strict mode before switching to Custom for some reason). I assume this is the dynamic isolation option in Custom.

You are correct.  That (Cross-site cookies) option is the one causing issues for cookie webextensions.  Cross-site tracking should still work as it was the previous firstparty isolation.

You should be able to select it even without going to strict first.  For advanced users there's the `about:config` page where you can do that as well.

**Comment by jouki2** [comment]
If the Cross-site cookies option is the one causing issues, why the same exact issue appears on both of my PCs, for like 1 month now. I was using the add-on with containers (strict mode) for years without a single problem. I tried Standard mode, Custom mode with whatever options (even with none of the  protection options ticked) and ~200 cookies just  cant be deleted, whatever I do.
I read that we have to wait for the issue to be fixed,  was just curious why Standart mode and no protection at all dont work as a current bugfix for me (on 2 PCs)

**Comment by kennethtran93** [comment]
> If the Cross-site cookies option is the one causing issues, why the same exact issue appears on both of my PCs, for like 1 month now. I was using the add-on with containers (strict mode) for years without a single problem. 
> I tried Standard mode, Custom mode with whatever options (even with none of the protection options ticked) and ~200 cookies just cant be deleted, whatever I do.
> I read that we have to wait for the issue to be fixed, was just curious why Standart mode and no protection at all dont work as a current bugfix for me (on 2 PCs)

They've adjusted the cookie protection in strict mode to cross-site protection in 86+.  This probably started setting newer and recently updated cookies to use an internal cookie value that we do not have access of.  

This is just my guess, but removal of cookies with cross-site protection requires that internal value that is not provided to us.  Changes to the protection would most likely only affect future cookies and not touch any existing ones.  This applies both ways, so cookies created while its behavior is cross-site protection will be ignored and newly created cookies using non-cross-site protection (can use cross-site tracking protection) should be removable by webextensions.

**Comment by rvalieris** [comment]
I had the same problem and I just did a test, changed from strict mode to custom mode and unchecked cookies, then I *manually deleted* all cookies using the `manage data...` button, and now it seems that new cookies are getting deleted as expected by the addon.

**Comment by kennethtran93** [comment]
@rvalieris Thank you for the tests and confirming the results. 

You may also use the 'cross-site tracking cookie' option as well. This was part of the 'strict' configuration up to and including 85 and should work with our cleanups.

**Comment by Dealz-416LS** [comment]
@kennethtran93 

Just to confirm, Strict Mode "Cross-site Cookies" (dFPI) gave me issues (cookies not being removed on cleanup), I switched to Custom Mode w/ "Cross-site tracking cookies" a while ago and not had a problem since.

**Comment by ghost** [comment]
Just reporting in: I'm running custom mode with everything turned on and it's given me no troubles until today. I had a lot of trouble with being half-logged-in to amazon prime video, suspected cookie issues, ran a clean, no fix, took me some time and eventually I manually cleaned the cookies from firefox's settings and everything came good. I can only conclude that the cleanup of cookies from the addon, didn't actually remove the cookies.

**Comment by kennethtran93** [comment]
> Just reporting in: I'm running custom mode with everything turned on and it's given me no troubles until today. I had a lot of trouble with being half-logged-in to amazon prime video, suspected cookie issues, ran a clean, no fix, took me some time and eventually I manually cleaned the cookies from firefox's settings and everything came good. I can only conclude that the cleanup of cookies from the addon, didn't actually remove the cookies.

You're right on that part. We had just assumed that the cookies that we requested to be deleted (that was fetched from the list of cookies returned) actually got deleted as it wasn't an issue until 86+.

Building a check is on my radar.

**Comment by kennethtran93** [comment]
An update from https://bugzilla.mozilla.org/show_bug.cgi?id=1669716 reveals that a patch to the extension API for 'partitionKey' (needed for cross-site protection buckets) is added for milestone Firefox 94.  Once this available on Dev channel, I'll plan some time to look into getting it working again.  I can't guarantee that I'll have it ready before Firefox 94 is released.

**Comment by klausenbusk** [comment]
The commit [has landed in mozilla-central](https://hg.mozilla.org/mozilla-central/rev/20b66bbd2a66) so should be in the next Nightly. The API is documented [here](https://phabricator.services.mozilla.com/D123491#4106552).

**Comment by Bobb-E-Burns** [comment]
FF94 landed as the main release this week and after updating I'm running into this issue where CAD does not want to actually delete anything.  
Has there been any progress with this issue with regards to FPI and custom (3rd party) blocking? @kennethtran93

**Comment by kennethtran93** [comment]
> FF94 landed as the main release this week and after updating I'm running into this issue where CAD does not want to actually delete anything.  
> Has there been any progress with this issue with regards to FPI and custom (3rd party) blocking? @kennethtran93 

The existing FPI (third party/cross-site tracking) option should still work. I'm working on the dynamic version (cross-site cookies). In the process of adding more info to the cookies so that you can select which partitioned cookie to keep/erase.

**Comment by Bobb-E-Burns** [comment]
> The existing FPI (third party/cross-site tracking) option should still work.  

CAD wasn't logging any automatic deletions yesterday (probably since earlier but only noticed it yesterday) and when looking at the storage through `about:preferences#privacy > Manage Data...` it confirmed nothing was being deleted.  
Then I found this issue and after reading through it (specifically comment [838437838](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/issues/807#issuecomment-838437838)) I fiddled with FF's cookie settings a bit and doing manual cleanups but it didn't help. Since I, likely, wrongly assumed my problem was related to this issue (due to your comment at [930830620](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/issues/807#issuecomment-930830620) specifically mentioning 94 and me just having updated to 94) I figured I'd wait for your reply and reverted to my original settings, but today, after restarting Firefox, it seems to behave (mostly) normally again. Not sure what was going on earlier.  

Only problem I see today is that, for some domains, the logger says it deleted "0 Cookie(s)" while it actually does delete everything, according to the `Manage Data...` dialog.

**Comment by kennethtran93** [comment]
> CAD wasn't logging any automatic deletions yesterday (probably since earlier but only noticed it yesterday) and when looking at the storage through `about:preferences#privacy > Manage Data...` it confirmed nothing was being deleted. Then I found this issue and after reading through it (specifically comment [838437838](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/issues/807#issuecomment-838437838)) I fiddled with FF's cookie settings a bit and doing manual cleanups but it didn't help. Since I, likely, wrongly assumed my problem was related to this issue (due to your comment at [930830620](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/issues/807#issuecomment-930830620) specifically mentioning 94 and me just having updated to 94) I figured I'd wait for your reply and reverted to my original settings, but today, after restarting Firefox, it seems to behave (mostly) normally again. Not sure what was going on earlier.

Just changing the privacy setting doesn't affect any existing cookies that was set with previous configurations.  That only applies to newly created cookies and/or newly visited sites.  And sometimes, a 'turn off and on again' would do the trick.


> Only problem I see today is that, for some domains, the logger says it deleted "0 Cookie(s)" while it actually does delete everything, according to the `Manage Data...` dialog.

The 'delete 0 cookies' usually means that other site data is cleaned up. If you expand the clean up history log (if you enabled it), it would tell you that other browsing data types are being cleaned (even if there are none).  This is because, unlike cookies, we cannot search through all the other browsing data types, so we manually set a cookie on the sites you visit so that there's at least one cookie in there that allows us to know the domain to trigger cleanup for.  The notification and logs will not account for the cookie that this webextension has set.

**Comment by Bobb-E-Burns** [comment]
> The 'delete 0 cookies' usually means that other site data is cleaned up.  

I have indeed checked everything for deletion, but FF's `Manage Data...` actually showed the domain had 1 cookie stored.   
So I just checked the cookie storage with FF's dev tools and the cookie is called `CookieAutoDeleteBrowsingDataCleanup` which let me to this page on your wiki  

https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/wiki/Documentation#other-browsing-data-cleanup-options

which I did not know about, and so that explains it. Although that document does erroneously state

> CAD will place a temporary cookie on all visited sites (that lasts for a few seconds) if you have the any of the other browsing data setting on.

It seems the cookie is actually just set on page load and it stays there the entire time until it is removed with cleanup (and correctly not counted as a cookie in the CAD logger, which explains the "0 cookies")  
Or am I misinterpreting it and it means that the cookie is set only if the visit lasts at least a few seconds?

**Comment by armaguedes** [comment]
Hey there.

For me, this problem has gone away from v94 (I'm currently on version 94.0.1x64, on Win10). I do not recall having changed any settings, one day it just went from "~598 cookies not deleted", to always deleting everything; the only difference I noticed was the new-Firefox tab thing.

**Comment by kennethtran93** [comment]
Just a heads up to all - it looks like since Firefox 94, the new dynamic partition / cross-site cookie / strict mode will not return any dynamically partitioned cookies through `getAll` without an additional attribute, so those cookies will still exist for the time being (CAD currently doesn't know that it exists because the browsers aren't returning them).  Previously, they were being returned but couldn't be deleted in any way (thus the constant similar cleanup logs).

On the bright side, the dynamic partition / strict mode / cross-site cookies should currently contain all third-party cookies to their specific first party sites, so many third-party sites should not easily know if you visited another site with their third-party cookies.

**Comment by lhindir** [comment]
Hi, any updates on this in the past few months? Thanks for your work!

**Comment by den7007** [comment]
Firefox now at 101.0.1, wondering if this is still an issue?

**Comment by thrnz** [comment]
'Total cookie protection' (ie. dfpi) is now rolling out by default, so this issue might begin effecting more users over the coming weeks/months.

https://blog.mozilla.org/en/products/firefox/firefox-rolls-out-total-cookie-protection-by-default-to-all-users-worldwide/

**Comment by wagnerand** [comment]
@kennethtran93 The Firefox docs have been updated which should hopefully enable you to fix this: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies#storage_partitioning

Most notably:
> By default, [cookies.get()](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/get), [cookies.getAll()](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/getAll), [cookies.set()](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/set), and [cookies.remove()](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/remove) work with cookies in unpartitioned storage. To work with cookies in partitioned storage in these APIs, topLevelSite in partitionKey must be set. **The exception is getAll where setting partitionKey without topLevelSite returned cookies in partitioned and unpartitioned storage**.

**Comment by ghost** [comment]
@kennethtran93 If possible would you mind giving an update about this issue ? Because everyday this addon becomes more useless which is sucks cause the addon itself is great.

**Comment by ghost** [comment]
I guess someone is pushing on bugzilla, what attention is mozilla paying? Any hope of being able to see this fixed in the short-medium term?

I view the current state of dFPI with concern, especially with [this other bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1767271).

**Comment by hbednar** [comment]
@kennethtran93 
It may be worth graying out the unsupported settings so that people don't keep opening issues related to this.
Or at least until there is a workaround, fix or firefox change.

**Comment by bege10** [comment]
@kennethtran93 I read the issues concerning the partitioned cookies and didn't understand if or when deleting these cookies will be supported by CAD.

**Comment by ghost** [comment]
> I guess someone is pushing on bugzilla, what attention is mozilla paying? Any hope of being able to see this fixed in the short-medium term?
> 
> I view the current state of dFPI with concern, especially with [this other bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1767271).

This bug you linked has been opened for 2 years and firefox still haven't fixed this. Idk what are they doing. 

CAD doesn't work and doing manual exception breaks partitioning (as mentioned in the linked bug) which is really frustrating.

They should make a universal and to the point permission manager (like bookmark manager) where you can granularly define each settings for how the site access/stores anything from/in the browser. 

Brave also has confusing settings imo. They both need to have a better way which is not a mess as it is currently.

**Comment by RalfJung** [comment]
> > I view the current state of dFPI with concern, especially with [this other bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1767271).
> 
> This bug you linked has been opened for 2 years and firefox still haven't fixed this. Idk what are they doing.

I don't see how that bug is related to this CAD issue? It's some other problem with dFPI.

The CAD issues was https://bugzilla.mozilla.org/show_bug.cgi?id=1669716 which has been fixed years ago, but apparently that was not enough. I presume it is [because of this](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/issues/807#issuecomment-974766508):

> Just a heads up to all - it looks like since Firefox 94, the new dynamic partition / cross-site cookie / strict mode will not return any dynamically partitioned cookies through getAll without an additional attribute, so those cookies will still exist for the time being (CAD currently doesn't know that it exists because the browsers aren't returning them). Previously, they were being returned but couldn't be deleted in any way (thus the constant similar cleanup logs).

So... does this need further Firefox fixes to expose partitioned cookies via `getAll`? Or does this need CAD to set some additional attribute on `getAll` to get the remaining cookies?

**Comment by mpan-pl** [comment]
It seems that CAD works as expected if privacy mode is set to “Custom” and, when cookies blocking is enabled, it’s switched away from “Isolate cross-site cookies” option. Perhaps that would point to a method to fix this issue?

For now that also works as a workaround too (tested Fx 148). But that’s suboptimal, in particular for users who don’t have other protections enabled.

**Comment by RalfJung** [comment]
Isolating cross-site cookies is way too important to disable it. It's one of the best advances in online privacy recently.

**Comment by mpan-pl** [comment]
@RalfJung, I feel like @kennethtran93 comparing CAD’s behavior with and without that option would rather improve online privacy than hurt it.

**Comment by RalfJung** [comment]
Without isolation, cross-site cookies leak information between sites via shared third-party domains. CAD doesn't prevent this if I have tabs open for different sites at the same time. So one has to choose between "I can be tracked across different sites" (prevented by site isolation) and "A single site can store cookies until I close my browser window" (prevented by CAD). Ideally we'd prevent both but that's not possible. Given the choice between those two options, to me it is clear -- I think it is much more important to prevent cross-site tracking, so I can't use CAD. Obviously everyone should make their own choice. But I find it strange that CAD recommends disabling cross-site isolation without a big fat disclaimer about the privacy issues this causes.

**Comment by mpan-pl** [comment]
@RalfJung, I’m not going to speak for CAD developer(s) and it’s up to them to make decisions, but I don’t think they’d be particularly concerned about leaking information between sites to compare CAD’s behavior. Even if, one can test it on sites under their own control.
